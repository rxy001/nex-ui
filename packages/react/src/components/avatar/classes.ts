import { generateUtilityClasses } from '../utils'

export const avatarClasses = generateUtilityClasses('nui-avatar', [
  'root',

  'radius-sm',
  'radius-md',
  'radius-lg',
  'radius-xl',
  'radius-full',
  'radius-none',

  // "blue" | "gray" | "pink" | "purple" | "cyan" | "yellow" | "orange" | "red" | "green"
  'color-gray',
  'color-green',
  'color-red',
  'color-yellow',
  'color-blue',
  'color-pink',
  'color-cyan',
  'color-purple',
  'color-orange',

  // "sm" | "md" | "lg"
  'size-sm',
  'size-md',
  'size-lg',

  'outlined',

  'img',
])

export const avatarGroupClasses = generateUtilityClasses('nui-avatar-group', [
  'root',
  'surplus',
])
