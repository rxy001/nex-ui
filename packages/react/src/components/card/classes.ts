import { generateUtilityClasses } from '../utils'

export const cardClasses = generateUtilityClasses('nui-card', [
  'root',
  'blurred',
  'hoverable',

  'shadow-xs',
  'shadow-sm',
  'shadow-md',
  'shadow-lg',
  'shadow-xl',

  'radius-none',
  'radius-sm',
  'radius-md',
  'radius-lg',
])

export const cardHeaderClasses = generateUtilityClasses('nui-card-header', [
  'root',
  'content',
  'title',
  'subtitle',
])

export const cardBodyClasses = generateUtilityClasses('nui-card-body', ['root'])

export const cardFooterClasses = generateUtilityClasses('nui-card-footer', [
  'root',
])

export const cardActionAreaClasses = generateUtilityClasses(
  'nui-card-action-area',
  ['root'],
)
