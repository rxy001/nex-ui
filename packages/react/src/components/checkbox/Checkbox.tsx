'use client'

import { cloneElement, isValidElement, useId, useMemo } from 'react'
import { isFunction, __DEV__, isString } from '@nex-ui/utils'
import { useControlledState, useEvent, useFocusRing } from '@nex-ui/hooks'
import { checkboxRecipe } from '../../theme/recipes'
import { useNexUI } from '../provider'
import { useCheckboxGroup } from './CheckboxGroupContext'
import {
  useDefaultProps,
  useStyles,
  composeClasses,
  getUtilityClass,
  useSlot,
} from '../utils'
import { CheckedIcon } from './CheckedIcon'
import { IndeterminateIcon } from './IndeterminateIcon'
import type { CSSObject } from '@emotion/react'
import type { CheckboxOwnerState, CheckboxProps } from './types'
import type {
  ElementType,
  ChangeEvent,
  HTMLAttributes,
  KeyboardEvent,
  InputHTMLAttributes,
  MouseEvent,
  ReactElement,
} from 'react'

const useSlotClasses = (ownerState: CheckboxOwnerState) => {
  const { prefix } = useNexUI()

  const { radius, size, color, disabled, checked, classes, indeterminate } =
    ownerState

  return useMemo(() => {
    const checkboxRoot = `${prefix}-checkbox`

    const slots = {
      root: [
        'root',
        `radius-${radius}`,
        `size-${size}`,
        `color-${color}`,
        disabled && 'disabled',
        checked && 'checked',
        indeterminate && 'indeterminate',
      ],
      input: ['input'],
      label: ['label'],
      icon: ['icon'],
    }

    return composeClasses(slots, getUtilityClass(checkboxRoot), classes)
  }, [checked, classes, color, disabled, indeterminate, prefix, radius, size])
}

const useSlotAriaProps = (
  ownerState: CheckboxOwnerState,
): Record<'input' | 'label', HTMLAttributes<HTMLElement>> => {
  const {
    disabled,
    type,
    checked,
    children,
    value,
    role,
    slotProps,
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledBy,
    'aria-checked': ariaChecked,
    'aria-disabled': ariaDisabled,
    as = 'input',
    tabIndex = 0,
  } = ownerState

  const id = useId()

  return useMemo(() => {
    const labelProps = slotProps?.label
    const stringChildren = isString(children)

    const label = {
      id: labelProps?.id ?? (stringChildren ? id : undefined),
    }

    let input: InputHTMLAttributes<HTMLInputElement> = {
      'aria-labelledby': ariaLabelledBy ?? label.id,
      'aria-label': ariaLabel ?? (stringChildren ? children : undefined),
      tabIndex: disabled ? -1 : tabIndex,
    }

    if (as === 'input' || isFunction(as)) {
      input = {
        disabled,
        checked,
        type,
        value,
        ...input,
      }
    } else {
      input = {
        role: role ?? 'checkbox',
        'aria-checked': ariaChecked ?? checked,
        'aria-disabled': ariaDisabled ?? (disabled || undefined),
        ...input,
      }
    }

    return { input, label }
  }, [
    ariaChecked,
    ariaDisabled,
    ariaLabel,
    ariaLabelledBy,
    as,
    checked,
    children,
    disabled,
    id,
    role,
    slotProps?.label,
    tabIndex,
    type,
    value,
  ])
}

export const Checkbox = <CheckboxComponent extends ElementType = 'input'>(
  inProps: CheckboxProps<CheckboxComponent>,
) => {
  const { primaryThemeColor, css } = useNexUI()

  const props = useDefaultProps<CheckboxProps>({
    name: 'Checkbox',
    props: inProps,
  })

  const groupCtx = useCheckboxGroup()

  const inGroup = !!groupCtx

  if (__DEV__ && inGroup) {
    if ('checked' in props) {
      console.warn(
        '[Nex UI] Checkbox: The CheckboxGroup is being used, `checked` will be ignored. Use the `value` of the CheckboxGroup instead.',
      )
    }
    if ('defaultChecked' in props) {
      console.warn(
        '[Nex UI] Checkbox: The CheckboxGroup is being used, `defaultChecked` will be ignored. Use the `defaultValue` of the CheckboxGroup instead.',
      )
    }

    if (!('value' in props)) {
      console.warn(
        '[Nex UI] Checkbox: The CheckboxGroup is being used, but `value` is not provided.',
      )
    }
  }

  const {
    sx,
    icon,
    value,
    className,
    children,
    slotProps,
    onCheckedChange,
    indeterminate,
    checked: checkedProp,
    type = 'checkbox',
    defaultChecked = false,
    name = groupCtx?.name,
    color = groupCtx?.color ?? primaryThemeColor,
    disabled = groupCtx?.disabled ?? false,
    size = groupCtx?.size ?? 'md',
    radius = groupCtx?.radius ?? groupCtx?.size ?? size,
    ...remainingProps
  } = props

  const [rawChecked, setRawChecked] = useControlledState(
    checkedProp,
    defaultChecked,
    onCheckedChange,
  )

  const { focusVisible, focusProps } = useFocusRing()

  const checked = inGroup ? groupCtx.isChecked(value) : rawChecked

  const ownerState: CheckboxOwnerState = {
    ...props,
    defaultChecked,
    type,
    name,
    disabled,
    color,
    checked,
    size,
    radius,
    inGroup,
  }

  const handleChange = useEvent((event: ChangeEvent<HTMLInputElement>) => {
    if (inGroup && value) {
      groupCtx.toggleValue(value)
    }

    if (!inGroup) {
      setRawChecked(event.target.checked)
    }
  })

  const handleClick = useEvent((event: MouseEvent<HTMLInputElement>) => {
    if (
      event.currentTarget.tagName !== 'INPUT' &&
      event.currentTarget === event.target
    ) {
      if (inGroup && value) {
        groupCtx.toggleValue(value)
      }

      if (!inGroup) {
        setRawChecked(!checked)
      }
    }
  })

  const handleKeyUp = useEvent((event: KeyboardEvent<HTMLInputElement>) => {
    // Keyboard accessibility for non interactive elements
    if (
      focusVisible &&
      event.key === 'Space' &&
      event.target === event.currentTarget &&
      event.currentTarget.tagName !== 'INPUT' &&
      event.currentTarget === event.target
    ) {
      event.currentTarget.click()
    }
  })

  const classes = useSlotClasses(ownerState)

  const styles = useStyles({
    ownerState,
    name: 'Checkbox',
    recipe: checkboxRecipe,
  })

  const slotAriaProps = useSlotAriaProps(ownerState)

  const [CheckboxRoot, getCheckboxRootProps] = useSlot({
    ownerState,
    elementType: 'label',
    externalSlotProps: slotProps?.root,
    style: styles.root,
    classNames: classes.root,
    additionalProps: {
      sx,
      className,
    },
  })

  const [CheckboxInput, getCheckboxInputProps] = useSlot({
    ownerState,
    elementType: 'input',
    externalForwardedProps: remainingProps,
    classNames: classes.input,
    style: styles.input,
    a11y: {
      ...slotAriaProps.input,
      onKeyUp: handleKeyUp,
    },
    additionalProps: {
      name,
      onChange: handleChange,
      onClick: handleClick,
      'data-focus-visible': focusVisible || undefined,
      ...focusProps,
    },
  })

  const [CheckboxIcon, getCheckboxIconProps] = useSlot({
    ownerState,
    elementType: 'span',
    externalSlotProps: slotProps?.icon,
    style: styles.icon,
    classNames: classes.icon,
  })

  const [CheckboxLabel, getCheckboxLabelProps] = useSlot({
    ownerState,
    elementType: 'span',
    externalSlotProps: slotProps?.label,
    style: styles.label,
    classNames: classes.label,
    a11y: slotAriaProps.label,
  })

  const renderCheckedIcon = () => {
    if (indeterminate) {
      return <IndeterminateIcon />
    }

    const customIcon = icon
      ? isFunction(icon)
        ? icon(ownerState)
        : icon
      : null

    if (!customIcon) {
      return <CheckedIcon checked={checked} />
    }

    if (isValidElement(customIcon)) {
      const element = customIcon as ReactElement<any>

      return cloneElement(element, {
        ...element.props,
        style: {
          ...element.props.style,
          ...(css(styles.checkedIcon) as CSSObject),
        },
      })
    }
    return customIcon
  }

  return (
    <CheckboxRoot {...getCheckboxRootProps()}>
      <CheckboxInput {...getCheckboxInputProps()} />
      <CheckboxIcon {...getCheckboxIconProps()}>
        {renderCheckedIcon()}
      </CheckboxIcon>
      {children && (
        <CheckboxLabel {...getCheckboxLabelProps()}>{children}</CheckboxLabel>
      )}
    </CheckboxRoot>
  )
}

Checkbox.displayName = 'Checkbox'
