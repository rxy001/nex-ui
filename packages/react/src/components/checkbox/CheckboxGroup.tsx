'use client'

import { useMemo } from 'react'
import { useControlledState } from '@nex-ui/hooks'
import { CheckboxGroupProvider } from './CheckboxGroupContext'
import type {
  CheckboxGroupContextValue,
  CheckboxGroupProps,
  CheckboxGroupValueType,
} from './types'

export const CheckboxGroup = <
  T extends CheckboxGroupValueType = CheckboxGroupValueType,
>(
  props: CheckboxGroupProps<T>,
) => {
  const {
    name,
    onValueChange,
    children,
    disabled,
    size,
    color,
    radius,
    defaultValue = [],
    value: valueProp,
  } = props

  const [values, setValues] = useControlledState(
    valueProp,
    defaultValue,
    onValueChange,
  )

  const ctx: CheckboxGroupContextValue<T> = useMemo(
    () => ({
      disabled,
      name,
      size,
      color,
      radius,
      toggleValue: (value: T) => {
        if (disabled) {
          return
        }

        let newValues: T[]
        if (values.includes(value)) {
          newValues = values.filter((existingValue) => existingValue !== value)
        } else {
          newValues = [...values, value]
        }

        setValues(newValues)
      },
      isChecked: (value?: T) => {
        return value ? values.includes(value) : false
      },
    }),
    [color, disabled, name, radius, setValues, size, values],
  )

  return (
    <CheckboxGroupProvider value={ctx as CheckboxGroupContextValue}>
      {children}
    </CheckboxGroupProvider>
  )
}

CheckboxGroup.displayName = 'CheckboxGroup'
