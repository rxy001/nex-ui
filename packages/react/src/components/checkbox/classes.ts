import { generateUtilityClasses } from '../utils'

// 用于单元测试
export const checkboxClasses = generateUtilityClasses('nui-checkbox', [
  // ---------root---------
  'root',

  // "sm" | "md" | "lg" | "full" | "none"
  'radius-sm',
  'radius-md',
  'radius-lg',
  'radius-full',
  'radius-none',

  // "blue" | "gray" | "pink" | "purple" | "cyan" | "yellow" | "orange" | "red" | "green"
  'color-gray',
  'color-green',
  'color-red',
  'color-yellow',
  'color-blue',
  'color-pink',
  'color-cyan',
  'color-purple',
  'color-orange',

  // "sm" | "md" | "lg"
  'size-sm',
  'size-md',
  'size-lg',

  'checked',
  'disabled',
  'indeterminate',
  // ---------root---------

  'input',
  'label',
  'icon',
])

export const checkboxGroupClasses = generateUtilityClasses(
  'nui-checkbox-group',
  [
    'root',
    'label',
    'wrapper',
    'orientation-vertical',
    'orientation-horizontal',
  ],
)
