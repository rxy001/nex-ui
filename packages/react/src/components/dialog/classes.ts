import { generateUtilityClasses } from '../utils'

export const dialogClasses = generateUtilityClasses('nui-dialog', [
  'root',
  'open',
  'backdrop',
])

export const dialogContentClasses = generateUtilityClasses(
  'nui-dialog-content',
  [
    'root',

    'size-xs',
    'size-sm',
    'size-md',
    'size-lg',
    'size-xl',
    'size-full',

    'full-screen',

    'placement-top',
    'placement-bottom',
    'placement-center',

    'scroll-inside',
    'scroll-outside',

    'paper',
    'close-button',
  ],
)

export const dialogHeaderClasses = generateUtilityClasses('nui-dialog-header', [
  'root',
])

export const dialogFooterClasses = generateUtilityClasses('nui-dialog-footer', [
  'root',
])

export const dialogBodyClasses = generateUtilityClasses('nui-dialog-body', [
  'root',
])
