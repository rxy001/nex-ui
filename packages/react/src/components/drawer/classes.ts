import { generateUtilityClasses } from '../utils'

export const drawerClasses = generateUtilityClasses('nui-drawer', [
  'root',
  'open',
  'backdrop',
])

export const drawerContentClasses = generateUtilityClasses(
  'nui-drawer-content',
  [
    'root',

    'placement-top',
    'placement-right',
    'placement-bottom',
    'placement-left',

    'size-xs',
    'size-sm',
    'size-md',
    'size-lg',
    'size-xl',
    'size-full',

    'paper',
    'close-button',
  ],
)

export const drawerHeaderClasses = generateUtilityClasses('nui-drawer-header', [
  'root',
])

export const drawerFooterClasses = generateUtilityClasses('nui-drawer-footer', [
  'root',
])

export const drawerBodyClasses = generateUtilityClasses('nui-drawer-body', [
  'root',
])
