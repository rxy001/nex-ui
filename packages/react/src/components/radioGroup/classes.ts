import { generateUtilityClasses } from '../utils'

// 用于单元测试
export const radioClasses = generateUtilityClasses('nui-radio', [
  // ---------root---------
  'root',

  // "blue" | "gray" | "pink" | "purple" | "cyan" | "yellow" | "orange" | "red" | "green"
  'color-gray',
  'color-green',
  'color-red',
  'color-yellow',
  'color-blue',
  'color-pink',
  'color-cyan',
  'color-purple',
  'color-orange',

  // "sm" | "md" | "lg"
  'size-sm',
  'size-md',
  'size-lg',

  'checked',
  'disabled',
  // ---------root---------

  'input',
  'label',
  'dot',
])

export const radioGroupClasses = generateUtilityClasses('nui-radio-group', [
  'root',
  'label',
  'wrapper',
  'orientation-vertical',
  'orientation-horizontal',
])
