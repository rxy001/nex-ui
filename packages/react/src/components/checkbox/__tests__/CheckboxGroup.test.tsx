import { useState } from 'react'
import {
  testComponentStability,
  renderWithNexUIProvider,
  testRootClassName,
  testRefForwarding,
  testClassNamesForwarding,
  testSlotPropsForwarding,
  testVariantDataAttrs,
} from '~/tests/shared'
import { fireEvent } from '@testing-library/react'
import { Checkbox } from '../Checkbox'
import { CheckboxGroup } from '../CheckboxGroup'
import { checkboxGroupClasses } from './classes'

const children = [
  <Checkbox value='orange' key='orange'>
    orange
  </Checkbox>,
  <Checkbox value='pear' key='pear'>
    pear
  </Checkbox>,
  <Checkbox value='apple' key='apple'>
    Apple
  </Checkbox>,
]

const slots = ['label', 'wrapper'] as const

describe('CheckboxGroup', () => {
  testComponentStability(<CheckboxGroup>{children}</CheckboxGroup>, {
    useAct: true,
  })

  testRootClassName(<CheckboxGroup>{children}</CheckboxGroup>, {
    useAct: true,
  })

  testVariantDataAttrs(
    <CheckboxGroup>{children}</CheckboxGroup>,
    ['orientation', ['vertical', 'horizontal']],
    {
      useAct: true,
    },
  )

  testRefForwarding(<CheckboxGroup />, {
    useAct: true,
  })

  testClassNamesForwarding(
    <CheckboxGroup label='Label'>{children}</CheckboxGroup>,
    slots,
    {
      label: 'test-label',
      wrapper: 'test-wrapper',
    },
    checkboxGroupClasses,
    {
      useAct: true,
    },
  )

  testSlotPropsForwarding(
    <CheckboxGroup label='Label'>{children}</CheckboxGroup>,
    slots,
    {
      label: {
        className: 'test-label',
      },
      wrapper: {
        className: 'test-wrapper',
      },
    },
    checkboxGroupClasses,
    {
      useAct: true,
    },
  )

  it('should render with default props', async () => {
    const { container } = await renderWithNexUIProvider(
      <CheckboxGroup>{children}</CheckboxGroup>,
      {
        useAct: true,
      },
    )
    const root = container.firstElementChild

    expect(root).toHaveClass(checkboxGroupClasses.root)
    expect(root).toHaveAttribute('data-orientation', 'horizontal')

    expect(root).toMatchSnapshot()
  })

  it('should render with label', async () => {
    const { rerender, queryByClassName } = await renderWithNexUIProvider(
      <CheckboxGroup>{children}</CheckboxGroup>,
      {
        useAct: true,
      },
    )

    expect(queryByClassName(checkboxGroupClasses.label)).not.toBeInTheDocument()

    rerender(<CheckboxGroup label='Fruits'>{children}</CheckboxGroup>)

    const label = queryByClassName(checkboxGroupClasses.label)
    expect(label).toBeInTheDocument()
    expect(label).toHaveTextContent('Fruits')
  })

  it('should switch different values when checkboxes are clicked', async () => {
    const { getAllByRole } = await renderWithNexUIProvider(
      <CheckboxGroup>{children}</CheckboxGroup>,
      {
        useAct: true,
      },
    )

    const [orangeCheckbox, pearCheckbox, appleCheckbox] =
      getAllByRole('checkbox')

    expect(orangeCheckbox).not.toBeChecked()
    expect(pearCheckbox).not.toBeChecked()
    expect(appleCheckbox).not.toBeChecked()

    fireEvent.click(orangeCheckbox)
    expect(orangeCheckbox).toBeChecked()

    fireEvent.click(pearCheckbox)
    expect(pearCheckbox).toBeChecked()

    fireEvent.click(appleCheckbox)
    expect(appleCheckbox).toBeChecked()
  })

  it('should be controlled by value prop', async () => {
    function ControlledCheckboxGroup() {
      const [value, setValue] = useState(['orange'])

      return (
        <CheckboxGroup value={value} onValueChange={setValue}>
          {children}
        </CheckboxGroup>
      )
    }

    const { getAllByRole } = await renderWithNexUIProvider(
      <ControlledCheckboxGroup />,
      {
        useAct: true,
      },
    )

    const [orangeCheckbox, pearCheckbox, appleCheckbox] =
      getAllByRole('checkbox')

    expect(orangeCheckbox).toBeChecked()
    expect(pearCheckbox).not.toBeChecked()
    expect(appleCheckbox).not.toBeChecked()

    fireEvent.click(pearCheckbox)
    expect(pearCheckbox).toBeChecked()
    expect(orangeCheckbox).toBeChecked()
    expect(appleCheckbox).not.toBeChecked()

    fireEvent.click(appleCheckbox)
    expect(appleCheckbox).toBeChecked()
    expect(orangeCheckbox).toBeChecked()
    expect(pearCheckbox).toBeChecked()

    fireEvent.click(orangeCheckbox)
    expect(orangeCheckbox).not.toBeChecked()
    expect(pearCheckbox).toBeChecked()
    expect(appleCheckbox).toBeChecked()
  })

  it('should support defaultValue prop', async () => {
    const { getAllByRole } = await renderWithNexUIProvider(
      <CheckboxGroup defaultValue={['orange']}>{children}</CheckboxGroup>,
      {
        useAct: true,
      },
    )

    const [orangeCheckbox, pearCheckbox, appleCheckbox] =
      getAllByRole('checkbox')

    expect(orangeCheckbox).toBeChecked()
    expect(pearCheckbox).not.toBeChecked()
    expect(appleCheckbox).not.toBeChecked()

    fireEvent.click(pearCheckbox)
    expect(pearCheckbox).toBeChecked()
    expect(orangeCheckbox).toBeChecked()
    expect(appleCheckbox).not.toBeChecked()

    fireEvent.click(appleCheckbox)
    expect(appleCheckbox).toBeChecked()
    expect(orangeCheckbox).toBeChecked()
    expect(pearCheckbox).toBeChecked()

    fireEvent.click(orangeCheckbox)
    expect(orangeCheckbox).not.toBeChecked()
    expect(pearCheckbox).toBeChecked()
    expect(appleCheckbox).toBeChecked()
  })

  it('should disable all checkboxes when disabled prop is true', async () => {
    const onValueChange = jest.fn()
    const { getAllByRole } = await renderWithNexUIProvider(
      <CheckboxGroup disabled onValueChange={onValueChange}>
        {children}
      </CheckboxGroup>,
      {
        useAct: true,
      },
    )

    const checkboxes = getAllByRole('checkbox')

    checkboxes.forEach((checkbox) => {
      expect(checkbox).toBeDisabled()
    })

    fireEvent.click(checkboxes[0])
    expect(onValueChange).not.toHaveBeenCalled()
  })

  it('should apply color, size, and radius props to checkboxes', async () => {
    const { getAllByRole } = await renderWithNexUIProvider(
      <CheckboxGroup color='green' size='lg' radius='lg'>
        {children}
      </CheckboxGroup>,
      {
        useAct: true,
      },
    )

    const checkboxes = getAllByRole('checkbox')

    checkboxes.forEach((checkbox) => {
      expect(checkbox.parentElement).toHaveAttribute('data-color', 'green')
      expect(checkbox.parentElement).toHaveAttribute('data-size', 'lg')
      expect(checkbox.parentElement).toHaveAttribute('data-radius', 'lg')
    })
  })

  it('should apply name prop to all checkboxes', async () => {
    const { getAllByRole } = await renderWithNexUIProvider(
      <CheckboxGroup name='fruits'>{children}</CheckboxGroup>,
      {
        useAct: true,
      },
    )

    const checkboxes = getAllByRole('checkbox')

    checkboxes.forEach((checkbox) => {
      expect(checkbox).toHaveAttribute('name', 'fruits')
    })
  })

  describe('Accessibility', () => {
    it('should have role="group" by default', async () => {
      const { getByTestId } = await renderWithNexUIProvider(
        <CheckboxGroup data-testid='group'>{children}</CheckboxGroup>,
        {
          useAct: true,
        },
      )
      expect(getByTestId('group')).toHaveAttribute('role', 'group')
    })

    it('should apply aria-labelledby to the root element when label is provided', async () => {
      const { getByTestId, getByRole, rerender } =
        await renderWithNexUIProvider(
          <CheckboxGroup data-testid='group'>{children}</CheckboxGroup>,
          {
            useAct: true,
          },
        )

      const group = getByTestId('group')
      expect(group).not.toHaveAttribute('aria-labelledby')

      rerender(
        <CheckboxGroup data-testid='group' label='Label'>
          {children}
        </CheckboxGroup>,
      )

      const heading = getByRole('heading', { name: 'Label' })
      expect(group).toHaveAttribute('aria-labelledby', heading.id)
    })
  })
})
