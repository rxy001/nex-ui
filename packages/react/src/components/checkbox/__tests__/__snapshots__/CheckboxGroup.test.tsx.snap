// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup should render with default props 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--nui-spaces-2);
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
  max-width: -webkit-fit-content;
  max-width: -moz-fit-content;
  max-width: fit-content;
  height: var(--nui-sizes-8);
  padding: var(--nui-spaces-1-5);
  -webkit-margin-start: calc(var(--nui-spaces-1-5) * -1);
  margin-inline-start: calc(var(--nui-spaces-1-5) * -1);
  -webkit-margin-end: var(--nui-spaces-1-5);
  margin-inline-end: var(--nui-spaces-1-5);
}

.emotion-3 {
  margin: 0;
  padding: 0;
  border: none;
  background: none;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  border-radius: 0;
  box-shadow: none;
  box-sizing: border-box;
  position: absolute;
  inset: 0;
  opacity: 0;
  cursor: inherit;
  z-index: 1;
}

.emotion-3[type="search"]::-webkit-search-decoration {
  -webkit-appearance: none;
}

.emotion-3[type="search"]::-webkit-search-cancel-button {
  -webkit-appearance: none;
}

.emotion-3:is(:-webkit-autofill, :autofill) {
  background-color: transparent!important;
  -webkit-transition: background-color 50000s ease-in-out 0s;
  transition: background-color 50000s ease-in-out 0s;
}

.emotion-3[data-focus-visible=true]:not(:disabled, [data-disabled=true])+*::before {
  outline: var(--nui-borders-focus-visible-outline);
  outline-offset: var(--nui-spaces-0-5);
}

.emotion-3:is(:hover, [data-hover=true]):not(:disabled, [data-disabled=true])+*::before {
  border-color: var(--nui-colors-blue-secondary);
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: var(--nui-colors-blue-contrast-text);
  width: var(--nui-sizes-5);
  height: var(--nui-sizes-5);
  font-size: var(--nui-font-sizes-md);
}

.emotion-4::before {
  content: "";
  position: absolute;
  inset: 0;
  border: var(--nui-borders-md);
  border-color: var(--nui-colors-gray-highlight);
  -webkit-transition: var(--nui-transitions-colors);
  transition: var(--nui-transitions-colors);
  background: transparent;
  z-index: -2;
  border-radius: calc(var(--nui-radii-lg) * .6);
}

.emotion-4::after {
  content: "";
  position: absolute;
  inset: 0;
  background-color: var(--nui-colors-blue-primary);
  -webkit-transition: all 0.2s linear;
  transition: all 0.2s linear;
  -webkit-transform: scale(0.4);
  -moz-transform: scale(0.4);
  -ms-transform: scale(0.4);
  transform: scale(0.4);
  opacity: 0;
  z-index: -1;
  border-radius: calc(var(--nui-radii-lg) * .6);
}

.emotion-5 {
  -webkit-margin-start: var(--nui-spaces-2);
  margin-inline-start: var(--nui-spaces-2);
  font-size: var(--nui-font-sizes-lg);
}

<div
  class="nui-checkbox-group-root nui-checkbox-group-orientation-horizontal emotion-0"
  role="group"
>
  <div
    class="nui-checkbox-group-wrapper emotion-1"
  >
    <label
      class="nui-checkbox-root nui-checkbox-radius-md nui-checkbox-size-md nui-checkbox-color-blue emotion-2"
    >
      <input
        aria-label="orange"
        aria-labelledby="«rh»"
        class="nui-checkbox-input emotion-3"
        tabindex="0"
        type="checkbox"
        value="orange"
      />
      <span
        class="nui-checkbox-icon emotion-4"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 12.9l3.143 3.6L18 7.5"
            opacity="0"
            pathLength="1"
            stroke-dasharray="0px 1px"
            stroke-dashoffset="0px"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
      <span
        class="nui-checkbox-label emotion-5"
        id="«rh»"
      >
        orange
      </span>
    </label>
    <label
      class="nui-checkbox-root nui-checkbox-radius-md nui-checkbox-size-md nui-checkbox-color-blue emotion-2"
    >
      <input
        aria-label="pear"
        aria-labelledby="«ri»"
        class="nui-checkbox-input emotion-3"
        tabindex="0"
        type="checkbox"
        value="pear"
      />
      <span
        class="nui-checkbox-icon emotion-4"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 12.9l3.143 3.6L18 7.5"
            opacity="0"
            pathLength="1"
            stroke-dasharray="0px 1px"
            stroke-dashoffset="0px"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
      <span
        class="nui-checkbox-label emotion-5"
        id="«ri»"
      >
        pear
      </span>
    </label>
    <label
      class="nui-checkbox-root nui-checkbox-radius-md nui-checkbox-size-md nui-checkbox-color-blue emotion-2"
    >
      <input
        aria-label="Apple"
        aria-labelledby="«rj»"
        class="nui-checkbox-input emotion-3"
        tabindex="0"
        type="checkbox"
        value="apple"
      />
      <span
        class="nui-checkbox-icon emotion-4"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m7 12.9l3.143 3.6L18 7.5"
            opacity="0"
            pathLength="1"
            stroke-dasharray="0px 1px"
            stroke-dashoffset="0px"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
      <span
        class="nui-checkbox-label emotion-5"
        id="«rj»"
      >
        Apple
      </span>
    </label>
  </div>
</div>
`;
